File: 백준/Gold/17822. 원판 돌리기/원판 돌리기.java
import java.io.*;
import java.util.*;

public class Main {

    public static int[] dr = {1, -1, 0, 0};
    public static int[] dc = {0, 0, 1, -1};
    public static int n, m, t;
    
    public static int[][] arr;
    public static int[] tmp;
    public static Deque<int[] > dq = new ArrayDeque<>();

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static StringTokenizer st;
    static boolean isContact(int r, int c, int num) {
        for (int i = 0; i < 4; ++i) {
            int nr = r + dr[i], nc = c + dc[i];
            if (nr < 0 || nr >= n) continue;
            nc = (nc + m)%m;
            if (arr[nr][nc] == num) return true;
        }
        return false;
    }

    static {
        n = 0; m = 0; t = 0;
    }

    public static int stoi (String s) {
        return Integer.parseInt(s);
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = stoi(st.nextToken());
        m = stoi(st.nextToken());
        t = stoi(st.nextToken());

        arr = new int[n][m];
        tmp = new int[m];

        for (int i = 0; i < n; ++i) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; ++j) {
                arr[i][j] = stoi(st.nextToken());
            }
        }

        int x, d ,k;

        while (t-- > 0) {
            st = new StringTokenizer(br.readLine());
            x = stoi(st.nextToken());
            d = stoi(st.nextToken()); 
            k = stoi(st.nextToken());
            

            // rotate
            for (int i = x-1; i < n; i += x) {
                for (int j = 0; j < m; ++j) {
                    tmp[((j + (d == 0 ? k : m-k))%m)] = arr[i][j];
                }
                System.arraycopy(tmp, 0, arr[i], 0, m);
            }

            boolean done =false;

            for (int r = 0; r < n; ++r) {
                for (int c = 0; c < m; ++c) {
                    if (arr[r][c] == 0) continue;
                    int score = arr[r][c];
                    if (!isContact(r, c, score)) continue;
                    done = true;
                    dq.add(new int[]{r, c});
                    arr[r][c] = 0;

                    while (!dq.isEmpty()) {
                        int[] loc = dq.poll();
                        int cr = loc[0], cc = loc[1];
                        for (int i = 0; i < 4; ++i) {
                            int nr = cr + dr[i], nc = cc + dc[i];
                            if (nr < 0 || nr >= n) continue;
                            nc = (nc + m)%m;
                            if (arr[nr][nc] != score) continue;
                            arr[nr][nc] = 0;
                            dq.push(new int[]{nr, nc});
                        }

                    }
                }
            }

            if (!done) {
                int cnt = 0, total = 0;
                for (int []line : arr) {
                    for (int a : line) {
                        if (a > 0) {
                            cnt++;
                            total+=a;
                        }
                    }
                }
                if (cnt == 0) break;
                double mean = 1.0 * total / cnt;
                for (int i = 0; i < n; ++i) {
                    for (int j = 0; j < m; ++j) {
                        if (arr[i][j] > 0) {
                            if (arr[i][j]*1.0 > mean) arr[i][j]--;
                            else if (arr[i][j] * 1.0 < mean) arr[i][j]++;
                        }
                    }
                }
            }
        }

        int ans = 0;
        for (int []ar : arr) {
            for (int a : ar) {
                ans+=a;
            }
        }
        System.out.println(ans);
        br.close();
    }
}

File: 백준/Gold/17822. 원판 돌리기/원판 돌리기.cc
#include <iostream>
#include <cstring>
#include <algorithm>
#include <vector>
#include <queue>
#include <deque>
using namespace std;
#define f(a, b, c) for(int a=b;a<c;++a)

int n, m, t;// 반지름, m개정수각각, t번 query
int offset_rotate[50]; // 0 초과,
deque<int> dq[50];
deque<pair<int ,int> > tmp;

int mv[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

void print() {
    f(i, 0, n) {
        f(j, 0, m) cout << dq[i][j] << ' ';
        cout << '\n';
    }

}

bool isContact (int r, int c) { //해당 위치 주변에 동일한 수가 있는가?
    int num = dq[r][c];
    for (auto &[dr, dc] : mv) {
        int nr = r + dr, nc = c + dc;
        if (nr < 0 || nr >= n) continue;
        nc = (nc + m) % m;
        if (num == dq[nr][nc]) return true;
    }
    return false;
}

int main(void) {
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    // freopen("input.txt", "r", stdin);
    cin >> n >> m >> t;
    f(i, 0, n) dq[i].resize(m);
    f(i, 0, n) f(j, 0, m) cin >> dq[i][j];
    int x, d, k;
    while (t--)
    {
        cin >> x >> d >> k; // Input

        if (k > m/2) {
            k = m-k;
            d ^= 1;
        }
        
        // rotate
        for (int i = x-1; i < n; i += x) {
            if (d) {
                f(j, 0, k) {
                    dq[i].push_back(dq[i].front());
                    dq[i].pop_front();
                }
            }
            else {
                f(j, 0, k) {
                    dq[i].push_front(dq[i].back());
                    dq[i].pop_back();
                }                
            }
        }

        // print();
        // cout << "---\n";
        //find 
        int done = false; // 
        f(r, 0, n) {
            f(c, 0, m) {
                bool found =false;
                if (!dq[r][c]) continue; // 0인경우 스킵
                if (!isContact(r, c)) continue; // 주변 동일한 수가 없음.


                // 있을 경우 주변 번호 전부 정리
                done = true;    
                int target_num = dq[r][c];
                tmp.push_back({r, c});
                dq[r][c] = 0;
                while (!tmp.empty())
                {
                    int cr = tmp.front().first, cc = tmp.front().second; tmp.pop_front();
                    for (auto &[dr, dc] : mv) {
                        int nr = cr + dr, nc = cc + dc;
                        if (nr < 0 || nr >= n) continue;
                        nc = (nc + m) % m;
                        if (dq[nr][nc] == target_num) {
                            dq[nr][nc] = 0;
                            tmp.push_back({nr, nc});
                        }
                    }
                }
                
            }
        }

        // 
        if (!done) {
            int cnt = 0, total = 0;
            f(i, 0, n) f(j, 0, m) {
                if (dq[i][j]) ++cnt;
                total += dq[i][j];
            }
            float mean = 1.0 * total / (1.0*cnt);
            f(i, 0, n) f(j, 0, m) {
                if (dq[i][j]) {
                    if (1.0 *dq[i][j] > mean) dq[i][j]--;
                    else if (1.0 *dq[i][j] < mean) dq[i][j]++;
                }
            }
        }
        // print();
        // cout <<"============\n";
    }

    int ans = 0;
    f(i, 0, n) {
        f(j, 0, m) {
            ans += dq[i][j];
        }
    }
    cout << ans;
    return 0;
}